File: AndroidDemoProject/app/src/main/java/com/by/androiddemoproject/router/RouteDefinitions.kt
package com.by.androiddemoproject.router

import android.net.Uri

object Routes {
    const val NATIVE_HOME = "app://native/home"
    const val NATIVE_DEMO = "app://native/demo"
    const val RN_HOME = "app://rn/home"
    const val RN_DETAIL = "app://rn/detail"
    const val FLUTTER_HOME = "app://flutter/home"
    const val FLUTTER_DETAIL = "app://flutter/detail"
    const val WEBVIEW = "app://webview"

    fun rnDetail(id: String) = "app://rn/detail?id=$id"
    fun flutterDetail(id: String) = "app://flutter/detail?id=$id"
    fun webview(url: String) = "app://webview?url=${Uri.encode(url)}"
}

fun Uri.queryParameters(): Map<String, String> {
    val params = mutableMapOf<String, String>()
    queryParameterNames.forEach { name ->
        getQueryParameter(name)?.let { params[name] = it }
    }
    return params
}

File: iOSDemoProject/iOSDemoProject/Router/RouteDefinitions.swift
import Foundation

struct Routes {
    static let nativeHome = "app://native/home"
    static let nativeDemo = "app://native/demo"
    static let rnHome = "app://rn/home"
    static let rnDetail = "app://rn/detail"
    static let flutterHome = "app://flutter/home"
    static let flutterDetail = "app://flutter/detail"
    static let webview = "app://webview"

    static func rnDetail(id: String) -> String {
        return "app://rn/detail?id=\(id)"
    }

    static func flutterDetail(id: String) -> String {
        return "app://flutter/detail?id=\(id)"
    }

    static func webview(url: String) -> String {
        return "app://webview?url=\(url.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? url)"
    }
}

extension URL {
    var queryParameters: [String: String] {
        guard let components = URLComponents(url: self, resolvingAgainstBaseURL: false),
              let queryItems = components.queryItems else {
            return [:]
        }
        var params: [String: String] = [:]
        for item in queryItems {
            params[item.name] = item.value ?? ""
        }
        return params
    }
}
